app/controllers/shopping_list_controller.rb

```diff
@@ -0,0 +1,27 @@
+class ShoppingListController < ApplicationController
+  def index
+    @foods = Food.includes(:user)
+      .where('foods.quantity < (SELECT COALESCE(SUM(recipe_foods.quantity), 0)
+        FROM recipe_foods WHERE recipe_foods.food_id = foods.id)')
+      .where(user_id: current_user.id).references(:recipe_foods)
+
+    @quantity_delta = Food.includes(:user)
+      .joins('LEFT JOIN recipe_foods ON recipe_foods.food_id = foods.id')
+      .group('foods.id')
+      .sum('COALESCE(recipe_foods.quantity, 0) - foods.quantity')
+
+    @prices = {}
+    @ingredients = @foods.each do |ingredient|
+      price_per_ingredient = if ingredient.quantity.zero?
+                               ingredient.price
+                             else
+                               ingredient.price / ingredient.quantity
+                             end
+      price = @quantity_delta[ingredient.id] * price_per_ingredient
+      @prices[ingredient.id] = price
```

Comment:
- [ ] It is good practice to always maintain lean Controllers i.e. less logic in the controllers. To apply the principle of a lean controller and improve the organization of your code, you can consider moving some of the logic from the ShoppingListController to the relevant models.  Here's an example of how you could refactor the code:

```ruby
class ShoppingListController < ApplicationController
  def index
    @foods = Food.with_quantity_deficiency(user: current_user)

    @quantity_delta = Food.quantity_delta(user: current_user)

    @prices = Food.calculate_prices(@quantity_delta)

    @total_price = @prices.values.sum
  end
end
```

In the Food model, you can define the following methods:

```ruby

class Food < ApplicationRecord
  # ...

  def self.with_quantity_deficiency(user:)
    includes(:user)
      .where('foods.quantity < (SELECT COALESCE(SUM(recipe_foods.quantity), 0)
        FROM recipe_foods WHERE recipe_foods.food_id = foods.id)')
      .where(user_id: user.id)
      .references(:recipe_foods)
  end

  def self.quantity_delta(user:)
    joins('LEFT JOIN recipe_foods ON recipe_foods.food_id = foods.id')
      .group('foods.id')
      .where(user_id: user.id)
      .sum('COALESCE(recipe_foods.quantity, 0) - foods.quantity')
  end

  def self.calculate_prices(quantity_delta)
    prices = {}

    Food.find(quantity_delta.keys).each do |ingredient|
      price_per_ingredient = if ingredient.quantity.zero?
                               ingredient.price
                             else
                               ingredient.price / ingredient.quantity
                             end

      price = quantity_delta[ingredient.id] * price_per_ingredient
      prices[ingredient.id] = price
    end

    prices
  end

  # ...
end
```

With this refactoring, the logic related to retrieving foods for the shopping list, calculating quantity deltas, and calculating prices is encapsulated within the Food model. The controller's responsibility is reduced to orchestrating the data retrieval and passing it to the view.

By following this approach, you adhere to the principle of a lean controller and also improve the maintainability and organization of your code.